sematic/graph.py

```diff
@@ -0,0 +1,544 @@
+        """
+        Within a given layer, get a run's downstream run IDs, depth-first.
+        """
+        output_edges: List[Edge] = self._edges_by_source_id.get(run_id, [])
+
+        downstream_ids: List[RunID] = []
+
+        for output_edge in output_edges:
+            if output_edge.destination_run_id is not None:
+                downstream_ids.append(output_edge.destination_run_id)
+                downstream_ids += self._get_run_downstream_ids(
+                    output_edge.destination_run_id
+                )
+
+        return list(set(downstream_ids))
+
+    def _get_skip_reset_run_ids(
+        self, reset_from: RunID
+    ) -> Tuple[List[RunID], List[RunID]]:
+        """
+        Figures out what run IDs to skip or reset based on rerun_from.
+
+        We skip descendants of all downstream of reset point plus descendants of
+        downstreams of ancestors. The skipped futures will be naturally
+        re-created by the new graph resolution.
+
+        reset = forcing future state to CREATED or RAN. Considering reset_from
+        and ancestors runs, we reset the run and all downstream.
+
+        Parameters
+        ----------
+        reset_from: RunID
+            ID of run from which to reset the graph
+
+        Returns
+        -------
+        Tuple[List[RunID], List[RunID]]
+            A tuple whose first element is the list of run IDs to skip when
+            cloning the graph. The second element is the list of run IDs whose
+            cloned future's state to reset to CREATED.
+        """
+        skip_run_ids: List[RunID] = []
+
+        reset_run_ids: List[RunID] = [reset_from]
+
+        ancestor_run_ids = self._get_run_ancestor_ids(reset_from)
+        reset_run_ids += ancestor_run_ids
+
+        for ancestor_run_id in [reset_from] + ancestor_run_ids:
+            downstream_run_ids = self._get_run_downstream_ids(ancestor_run_id)
+            reset_run_ids += downstream_run_ids
+
+            for downstream_run_id in downstream_run_ids:
+                downstream_descendant_ids = self._get_run_descendant_ids(
+                    downstream_run_id
+                )
+                skip_run_ids += downstream_descendant_ids
+
+        return skip_run_ids, reset_run_ids
+
+    @memoized_indexed
+    def _get_artifact_value(self, artifact_id: str) -> Any:
+        artifact = self._artifacts_by_id[artifact_id]
+        return get_artifact_value(artifact, self.storage)
+
+    def _get_cloned_future_inputs(
+        self, run_id: RunID, cloned_graph: ClonedFutureGraph
+    ) -> Tuple[Dict[str, Any], Dict[str, Artifact]]:
+        kwargs: Dict[str, Any] = {}
+
+        input_edges = self._edges_by_destination_id[run_id]
+
+        run_input_artifacts: Dict[str, Artifact] = {}
+
+        for input_edge in input_edges:
+            if input_edge.destination_name is None:
+                raise RuntimeError("Input edge misses destination name")
```

Comment:
When would these illegal state conditions ever happen?
If they do happen, then the error will be bubbled up to the user when they want to rerun a graph, so the final error message should explicitly state the fact that there is a problem with the run graph.

Reply:
This situation should never happen, this is a safeguard to help debugging.